RFC.md

```diff
@@ -0,0 +1,466 @@
+
+**TLDR**
+- Multimodal models are primarily DeepFusion and EarlyFusion architectures
+- Modules/utils for fusing and un-fusing MM models can allow for fine-tuning a variety of MM architectures in a uniform way.
+
+## Proposal
+
+**High Level List:**
+- Introduce a modules/model_fusion folder for handling fused models
+- fusion_models introduces two top level wrapper modules that combine a single encoder and decoder in two different ways. This allows the two models to be used as one model.
+- fusion_layer and fusion_embed are fusion modules (similar to PeFT modules) that introduce extra trainable parameters to the pre-trained fused models
+- fusion_utils is based on peft_utils and provides functions for accessing and modifying fusion parameters
+
+**Folder Structure for new modules/functions**
+```
+torchtune/
+	modules/
+		model_fusion/
+			fusion_models.py
+				DeepFusionModel
+				EarlyFusionModel
+			fusion_layer.py
+			fusion_embed.py
+			fusion_utils.py
+				register_fusion_module
+				get_fusion_params
+				set_trainable_params
+```
+
+**Pseudo code demonstrating usage of fused models**
+```python
+def flamingo_builder():
+	return DeepFusionModel(
+		encoder=CLIP(...),
+		decoder=Llama3(...))
+
+>>> model = flamingo_builder()
+
+# Load full checkpoints
+>>> model.load_state_dict(...)
+
+# Or load pretrained individual models
+>>> model.encoder.load_state_dict(...)
+>>> model.decoder.load_state_dict(...)
+
+# set only the fusion parameters to be trainable
+>>> fusion_params = get_fusion_params(model)
+>>> set_trainable_params(model, fusion_params)
+
+# Do inference with the model the same as a text model
+# Fusion models have the same signatures as TransformerDeocder
+>>> data = flamingo_transform(...) # Message formated text + images
+>>> model(**data)
+tensor([-1.0885, ..., .0231], grad_fn=<ViewBackward0>)
+```
+
+> [!Note]
+> This RFC proposes a complete design for supporting fusion models, but not all of the design is required to be built to support any one given fusion style model. This RFC gives us an opportunity to take a holistic view over the space, but it is recommended that each module/component is built as need.
+### Fusion Models
+DeepFusion requires that Decoder is already modified to have any additional layers for accepting encoder inputs (these should be wrapped with FusionLayer to be able to track these additional parameters and work with checkpoints). For inference all of the cache can be managed by the decoder, so these methods are simply passed through.
+
+```python
+class DeepFusionModel(nn.Module):
+	def __init__(self,
+		decoder: TransformerDecoder,
+		encoder: nn.Module,
```

Comment:
Assuming we should have a ```TransformerEncoder``` class?

Reply:
I think we might want this in the future but I don't see any benefit to do this yet. TransformerDecoder is necessary because it gives us a contract that we use across our recipes as well as for decoding in generation and eval. So far we don't need this for the encoder so I'd rather leave it open as long as possible.

Reply:
You do have an implicit contract between the output of the encoder and the cross-attention layer? How should I reconcile?

Reply:
From a code perspective, I just need a Module where `out: Tensor = Module(**inputs)`. Now obviously the shape of out needs to match what you're encoder is expecting, but that's on the user to use the right encoder with the right decoder.